src/Microsoft.OData.Client/Serialization/ODataPropertyConverter.cs

```diff
@@ -664,6 +900,26 @@ private ODataResourceSetWrapper CreateODataComplexCollectionPropertyResourceSet(
             return this.CreateODataResourceSetWrapperForComplexCollection(property.PrimitiveOrComplexCollectionItemType, property.PropertyName, propertyValue, visitedComplexTypeObjects, isDynamic);
         }
 
+        /// <summary>
+        /// Returns the resource set of the collection dynamic property.
+        /// </summary>
+        /// <param name="propertyName">Name of the collection dynamic property.</param>
+        /// <param name="propertyValue">Collection instance.</param>
+        /// <param name="serverTypeName">The server type name of the entity whose properties are being populated.</param>
+        /// <param name="visitedComplexTypeObjects">List of instances of complex types encountered in the hierarchy. Used to detect cycles.</param>
+        /// <returns>An instance of ODataResourceSetWrapper representing the value of the property.</returns>
+        private ODataResourceSetWrapper CreateODataComplexCollectionDynamicPropertyResourceSet(string propertyName, object propertyValue, string serverTypeName, HashSet<object> visitedComplexTypeObjects)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(propertyName), "!string.IsNullOrEmpty(propertyName)");
+            Debug.Assert(propertyValue != null, "propertyValue != null");
+            Debug.Assert(propertyValue is ICollection, "propertyValue must be a collection");

```

Comment:
Same case for these. Could we check this in release mode as well?

Reply:
@paulodero It's common practice to apply asserts in debug mode and not release mode for performance reasons - though in Release you'd typically use `Trace.Assert()`. Assertions will help you verify your assumptions during development.  An assertion in Release mode can't save you. Your program would crash either way, same way it'll crash if without the assert.
`Debug.Assert(customer != null)` will yield the same outcome as try to access a property from the null reference - `customer.Name` in Release mode. In Debug you can do something about it

Reply:
Now that we know the object can be null, can the function handle this to ensure the program does not crash?